# Assignment 1: Implementing Document Loaders in LangChain

## Objective:
Write a Python script that uses LangChain’s document loaders to load documents from a directory. Your task is to implement functionality that reads `.txt` and `.pdf` files and outputs their content as LangChain `Document` objects.

---

## Requirements:
1. Use LangChain’s `TextLoader` for `.txt` files and `PyPDFLoader` for `.pdf` files.
2. Implement a function `load_documents(directory: str)` that:
   - Iterates through all files in the specified directory.
   - Loads the content of `.txt` and `.pdf` files.
   - Returns a list of `Document` objects, where:
     - **Page Content:** Contains the file’s text content.
     - **Metadata:** Includes the filename or other relevant file metadata.
3. Handle unsupported file types or errors gracefully.

---

## Input:
- A directory containing files with extensions `.txt` and `.pdf`.

---

## Output:
- A list of LangChain `Document` objects. Each document should contain:
  - The text content of the file.
  - Metadata such as the filename.

---

## Example:
### Input:
A directory with the following files:
- `example.txt` containing "Hello, this is a text file."
- `example.pdf` containing "This is a PDF document."
- `image.jpg` (unsupported file type).

### Output:
```python
[
    Document(page_content="Hello, this is a text file.", metadata={"filename": "example.txt"}),
    Document(page_content="This is a PDF document.", metadata={"filename": "example.pdf"})
]

In [5]:
from langchain.document_loaders import TextLoader, PyPDFLoader
from langchain.schema import Document
import os
from typing import List

def load_documents(directory: str) -> List[Document]:
    """
    Skeleton Function: Load .txt and .pdf documents from a directory.

    Args:
        directory (str): Path to the directory containing files.

    Returns:
        List[Document]: A list of LangChain Document objects.
    """
    # Initialize an empty list to store the documents
    documents = []

    # Loop through files in the specified directory
    for filename in os.listdir(directory):
        file_path = os.path.join(directory, filename)

        try:
            # Placeholder for loading .txt files
            if filename.endswith(".txt"):
                loader = TextLoader(file_path)
                documents.extend(loader.load())
            # Placeholder for loading .pdf files
            elif filename.endswith(".pdf"):
                loader = PyPDFLoader(file_path)
                documents.extend(loader.load())

        except Exception as e:
            # Print error for files that could not be loaded
            print(f"Error loading {filename}: {e}")

    # Return the list of documents
    return documents

# Example usage
if __name__ == "__main__":
    # Define the directory path
    directory_path = "/home/kayozxo/Documents/techfest/assignment"

    # Call the function to load documents
    docs = load_documents(directory_path)

    # Iterate through the loaded documents and print metadata and content preview
    for doc in docs:
        print(f"File: {doc.metadata.get('filename', 'Unknown')}, Content Preview: {doc.page_content[:100]}")

File: Unknown, Content Preview: Provided proper attribution is provided, Google hereby grants permission to
reproduce the tables and
File: Unknown, Content Preview: 1 Introduction
Recurrent neural networks, long short-term memory [13] and gated recurrent [7] neural
File: Unknown, Content Preview: Figure 1: The Transformer - model architecture.
The Transformer follows this overall architecture us
File: Unknown, Content Preview: Scaled Dot-Product Attention
 Multi-Head Attention
Figure 2: (left) Scaled Dot-Product Attention. (r
File: Unknown, Content Preview: output values. These are concatenated and once again projected, resulting in the final values, as
de
File: Unknown, Content Preview: Table 1: Maximum path lengths, per-layer complexity and minimum number of sequential operations
for 
File: Unknown, Content Preview: length n is smaller than the representation dimensionality d, which is most often the case with
sent
File: Unknown, Content Preview: Table 2: The Transformer achieves bet

File: Unknown, Content Preview: Provided proper attribution is provided, Google hereby grants permission to
reproduce the tables and
File: Unknown, Content Preview: 1 Introduction
Recurrent neural networks, long short-term memory [13] and gated recurrent [7] neural
File: Unknown, Content Preview: Figure 1: The Transformer - model architecture.
The Transformer follows this overall architecture us
File: Unknown, Content Preview: Scaled Dot-Product Attention
 Multi-Head Attention
Figure 2: (left) Scaled Dot-Product Attention. (r
File: Unknown, Content Preview: output values. These are concatenated and once again projected, resulting in the final values, as
de
File: Unknown, Content Preview: Table 1: Maximum path lengths, per-layer complexity and minimum number of sequential operations
for 
File: Unknown, Content Preview: length n is smaller than the representation dimensionality d, which is most often the case with
sent
File: Unknown, Content Preview: Table 2: The Transformer achieves better BLEU scores than previous state-of-the-art models on the
En
File: Unknown, Content Preview: Table 3: Variations on the Transformer architecture. Unlisted values are identical to those of the b
File: Unknown, Content Preview: Table 4: The Transformer generalizes well to English constituency parsing (Results are on Section 23
File: Unknown, Content Preview: [5] Kyunghyun Cho, Bart van Merrienboer, Caglar Gulcehre, Fethi Bougares, Holger Schwenk,
and Yoshua
File: Unknown, Content Preview: [25] Mitchell P Marcus, Mary Ann Marcinkiewicz, and Beatrice Santorini. Building a large annotated
c
File: Unknown, Content Preview: Attention Visualizations
Input-Input Layer5
...
is
what
File: Unknown, Content Preview: The world must be made safe for democracy. Its peace must be planted upon the tested
foundations of 
Output is truncated. View as a scrollable element or open in a text editor. Adjust cell output settings...

# Assignment: Chunking Data and Converting It to Vector Embeddings

## Objective:
Write a Python script that uses LangChain to:
1. Load `.txt` and `.pdf` files as `Document` objects from a directory.
2. Chunk the data into smaller pieces for efficient processing.
3. Convert the chunks into vector embeddings using a text embedding model.

---

## Requirements:
1. **Document Loading**:
   - Use LangChain’s `TextLoader` for `.txt` files and `PyPDFLoader` for `.pdf` files.
   - Implement a function `load_documents(directory: str)` to load all files from a directory as LangChain `Document` objects.

2. **Chunking**:
   - Use LangChain’s `RecursiveCharacterTextSplitter` to split the document text into smaller chunks.
   - Implement a function `chunk_documents(documents: List[Document]) -> List[Document]`.

3. **Embedding Generation**:
   - Use a pre-trained embedding model (e.g., `OpenAIEmbeddings` or any other LangChain-compatible embedding model).
   - Implement a function `generate_embeddings(chunks: List[Document]) -> List[List[float]]` that converts each chunk into a vector embedding.

4. **Error Handling**:
   - Handle unsupported file types and errors gracefully.

---

## Input:
- A directory containing `.txt` and `.pdf` files.

---

## Output:
- A list of vector embeddings for the chunks of the loaded documents.

---

## Example:
### Input:
A directory with the following files:
- `example.txt` containing "This is an example text file."
- `example.pdf` containing "This is an example PDF document."

### Output:
A list of embeddings (e.g., 768-dimensional vectors) for the chunks generated from the documents.

---

In [2]:
from langchain.schema import Document
from langchain.text_splitter import RecursiveCharacterTextSplitter
from sentence_transformers import SentenceTransformer
from typing import List

def chunk_documents(documents: List[Document]) -> List[Document]:
    """
    Splits documents into smaller chunks.

    Args:
        documents (List[Document]): List of LangChain Document objects.

    Returns:
        List[Document]: A list of chunked Document objects.
    """
    # Create an instance of the text splitter with specified chunk size and overlap
    text_splitter = RecursiveCharacterTextSplitter(chunk_size=512, chunk_overlap=50)
    chunks = []

    # Iterate over each document and split it into chunks
    for doc in documents:
        # Split the document and add chunks to the list
        chunks.extend(text_splitter.split_documents([doc]))

    return chunks

def generate_embeddings(chunks: List[Document]) -> List[List[float]]:
    """
    Generates vector embeddings for the given chunks.

    Args:
        chunks (List[Document]): List of chunked Document objects.

    Returns:
        List[List[float]]: A list of vector embeddings.
    """
    # Initialize the OpenAI embeddings model
    model = SentenceTransformer('all-MiniLM-L6-v2')

    # Generate embeddings for each chunk
    return [model.encode(chunk.page_content) for chunk in chunks]

# Example Usage
if __name__ == "__main__":
    # Sample directory path where documents are stored
    directory_path = "/home/kayozxo/Documents/techfest/assignment"

    # Load documents (This function is implemented in Assignment 1)
    documents = load_documents(directory_path)

    # Chunk the documents into smaller chunks
    chunks = chunk_documents(documents)

    # Generate embeddings for the chunks
    embeddings = generate_embeddings(chunks)

    # Display first 5 embeddings for demonstration
    for i, embedding in enumerate(embeddings[:5]):  # Display first 5 embeddings for brevity
        print(f"Embedding {i + 1}: {embedding[:10]}...")  # Print first 10 dimensions for brevity

/home/kayozxo/.local/lib/python3.13/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


Embedding 1: [ 0.11234583 -0.00280949  0.00497708  0.05177931  0.00560877 -0.00149715
 -0.03321598  0.00816553 -0.02711167  0.1374356 ]...
Embedding 2: [-0.1182005  -0.09273406  0.0547741  -0.0446618   0.01874045  0.05638526
 -0.09028511  0.0138909   0.05966606 -0.0552347 ]...
Embedding 3: [-0.04555681 -0.09249639  0.03179702  0.00627355 -0.01106313  0.00122013
 -0.00635648  0.00280261  0.10205809 -0.0459167 ]...
Embedding 4: [-0.04007926 -0.10125792  0.03301181  0.00473507 -0.00257197  0.0668017
 -0.03224802  0.05358506  0.00947471 -0.0510125 ]...
Embedding 5: [-0.08945089 -0.01966043 -0.0573628  -0.09178941 -0.03269776 -0.01818741
 -0.04241158  0.08523312  0.01385009  0.00655173]...


/home/kayozxo/.local/lib/python3.13/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
Embedding 1: [ 0.11234583 -0.00280949  0.00497708  0.05177931  0.00560877 -0.00149715
 -0.03321598  0.00816553 -0.02711167  0.1374356 ]...
Embedding 2: [-0.1182005  -0.09273406  0.0547741  -0.0446618   0.01874045  0.05638526
 -0.09028511  0.0138909   0.05966606 -0.0552347 ]...
Embedding 3: [-0.04555681 -0.09249639  0.03179702  0.00627355 -0.01106313  0.00122013
 -0.00635648  0.00280261  0.10205809 -0.0459167 ]...
Embedding 4: [-0.04007926 -0.10125792  0.03301181  0.00473507 -0.00257197  0.0668017
 -0.03224802  0.05358506  0.00947471 -0.0510125 ]...
Embedding 5: [-0.08945089 -0.01966043 -0.0573628  -0.09178941 -0.03269776 -0.01818741
 -0.04241158  0.08523312  0.01385009  0.00655173]...

# Assignment 3: Creating and Querying a Vector Database with Chroma

## Objective:
Write a Python script to:
1. Create a vector database using the **Chroma** library.
2. Store vector embeddings of document chunks in the database.
3. Query the database using similarity search and retrieve the top `k` results.

---

## Requirements:
1. **Vector Database Creation**:
   - Use Chroma to create a persistent vector database.
   - Add document embeddings (e.g., from OpenAI or any other embedding model) along with metadata to the database.

2. **Similarity Search**:
   - Implement a function to query the database with a user-provided text and retrieve the top `k` most similar results.

3. **Input Data**:
   - Use a list of text chunks or embeddings for this task. You may generate these from documents (e.g., `.txt` or `.pdf` files).

4. **Outputs**:
   - Return the metadata and content of the top `k` most similar results from the database.

---

## Example:
### Input:
1. A collection of text chunks from documents such as:
   - `"LangChain is a framework for developing applications powered by LLMs."`
   - `"Chroma is a vector database used for storing embeddings and performing similarity search."`
   - `"Document loaders are part of LangChain and help load data from multiple formats."`

2. Query text: `"What is Chroma?"`
3. `k=2`

### Output:
Top `k` results based on similarity:
1. Content: `"Chroma is a vector database used for storing embeddings and performing similarity search."`
   Metadata: `{...}`
2. Content: `"LangChain is a framework for developing applications powered by LLMs."`
   Metadata: `{...}`

---

In [4]:
from langchain.vectorstores import FAISS
from langchain_ollama import OllamaEmbeddings
from langchain.schema import Document
from typing import List
import os


def load_documents(directory_path: str) -> List[Document]:
    """
    Loads text files from a directory and creates LangChain Document objects.

    Args:
        directory_path (str): Path to the directory containing text files.

    Returns:
        List[Document]: A list of Document objects.
    """
    documents = []
    for filename in os.listdir(directory_path):
        file_path = os.path.join(directory_path, filename)
        if os.path.isfile(file_path) and filename.endswith(".txt"):
            with open(file_path, 'r', encoding='utf-8') as file:
                content = file.read()
                documents.append(Document(page_content=content, metadata={"filename": filename}))
    return documents


def initialize_faiss_with_ollama(documents: List[Document], db_path: str) -> FAISS:
    """
    Initializes a FAISS vector database with Ollama embeddings and stores documents.

    Args:
        documents (List[Document]): List of LangChain Document objects.
        db_path (str): Path to store the FAISS database.

    Returns:
        FAISS: FAISS vector store object.
    """
    # Initialize Ollama embeddings
    ollama_embeddings = OllamaEmbeddings(model="llama2")

    # Create FAISS vector store and store documents
    faiss_db = FAISS.from_documents(documents, ollama_embeddings)

    # Save the FAISS index to the specified path
    faiss_db.save_local(db_path)

    return faiss_db


def query_faiss(query: str, faiss_db: FAISS, k: int) -> List[str]:
    """
    Queries the FAISS database for the top-k similar documents.

    Args:
        query (str): Query text.
        faiss_db (FAISS): FAISS vector store object.
        k (int): Number of top results to return.

    Returns:
        List[str]: List of top-k document contents from the database.
    """
    # Perform similarity search in the database
    retrieved_results = faiss_db.similarity_search(query, k=k)

    # Collect the content of the retrieved documents
    results = [result.page_content for result in retrieved_results]

    return results


# Example Usage
if __name__ == "__main__":
    directory_path = "/home/kayozxo/Documents/techfest/assignment"

    # Load documents
    documents = load_documents(directory_path)

    # Initialize FAISS with Ollama embeddings
    db_path = "/home/kayozxo/Documents/techfest/assignment/faiss_index"
    faiss_db = initialize_faiss_with_ollama(documents, db_path)

    # Query FAISS database
    query_text = "What is FAISS?"
    top_k = 2
    results = query_faiss(query_text, faiss_db, top_k)

    for i, result in enumerate(results):
        print(f"Result {i+1}: {result}")


Result 1: The world must be made safe for democracy. Its peace must be planted upon the tested
foundations of political liberty. We have no selfish ends to serve. We desire no conquest, no
dominion. We seek no indemnities for ourselves, no material compensation for the sacrifices we
shall freely make. We are but one of the champions of the rights of mankind. We shall be
satisfied when those rights have been made as secure as the faith and the freedom of nations
can make them. Just because we fight without rancor and without selfish object, seeking nothing
for ourselves but what we shall wish to share with all free peoples, we shall, I feel confident,
conduct our operations as belligerents without passion and ourselves observe with proud
punctilio the principles of right and of fair play we profess to be fighting for. … It will be all the
easier for us to conduct ourselves as belligerents in a high spirit of right and fairness because
we act without animus, not in enmity toward a people

Result 1: The world must be made safe for democracy. Its peace must be planted upon the tested
foundations of political liberty. We have no selfish ends to serve. We desire no conquest, no
dominion. We seek no indemnities for ourselves, no material compensation for the sacrifices we
shall freely make. We are but one of the champions of the rights of mankind. We shall be
satisfied when those rights have been made as secure as the faith and the freedom of nations
can make them. Just because we fight without rancor and without selfish object, seeking nothing
for ourselves but what we shall wish to share with all free peoples, we shall, I feel confident,
conduct our operations as belligerents without passion and ourselves observe with proud
punctilio the principles of right and of fair play we profess to be fighting for. … It will be all the
easier for us to conduct ourselves as belligerents in a high spirit of right and fairness because
we act without animus, not in enmity toward a people or with the desire to bring any injury or
disadvantage upon them, but only in armed opposition to an irresponsible government which
has thrown aside all considerations of humanity and of right and is running amuck. We are, let
me say again, the sincere friends of the German people, and shall desire nothing so much as
the early reestablishment of intimate relations of mutual advantage between us—however hard
it may be for them, for the time being, to believe that this is spoken from our hearts. We have
borne with their present government through all these bitter months because of that friendship—
exercising a patience and forbearance which would otherwise have been impossible. We shall,
happily, still have an opportunity to prove that friendship in our daily attitude and actions toward
the millions of men and women of German birth and native sympathy who live among us and
share our life, and we shall be proud to prove it toward all who are in fact loyal to their neighbors
and to the government in the hour of test. They are, most of them, as true and loyal Americans
as if they had never known any other fealty or allegiance. They will be prompt to stand with us in
rebuking and restraining the few who may be of a different mind and purpose. If there should be
disloyalty, it will be dealt with with a firm hand of stern repression; but, if it lifts its head at all, it
...
itself at last free. To such a task we can dedicate our lives and our fortunes, everything that we
are and everything that we have, with the pride of those who know that the day has come when
America is privileged to spend her blood and her might for the principles that gave her birth and
happiness and the peace which she has treasured. God helping her, she can do no other.
Output is truncated. View as a scrollable element or open in a text editor. Adjust cell output settings...